Staging/EnableGop/EnableGop.c

```diff
@@ -0,0 +1,233 @@
+/** @file
+  Provide GOP on unsupported graphics cards on classic MacPro.
+
+  Copyright (c) 2022-2023, Mike Beaton. All rights reserved.<BR>
+  SPDX-License-Identifier: BSD-3-Clause
+**/
+
+#include <Uefi.h>
+
+#include <Library/DuetBdsLib.h>
+#include <Library/DxeServicesTableLib.h>
+#include <Library/OcBootManagementLib.h>
+#include <Library/OcConsoleLib.h>
+#include <Library/OcDeviceMiscLib.h>
+#include <Library/UefiRuntimeServicesTableLib.h>
+
+#include <Library/OcVariableLib.h>
+
+#define EFI_BLUETOOTH_DELAY_DEFAULT  3000
+
+STATIC EFI_GET_MEMORY_SPACE_MAP  mOriginalGetMemorySpaceMap = NULL;
+
+STATIC
+VOID
+LogStage (
+  CHAR8  Stage
+  )
+{
+  UINT8   One    = 1;
+  CHAR16  Name[] = L"efi-gop-stage0";
+
+  if (FeaturePcdGet (PcdEnableGopDirect)) {
+    Name[6] = L'd';
+  }
+
+  Name[L_STR_LEN (Name) - 1] = Stage;
+  OcSetSystemVariable (
+    Name,
+    OPEN_CORE_NVRAM_ATTR,
+    sizeof (One),
+    &One,
+    NULL
+    );
+}
+
+STATIC
+EFI_STATUS
+LoadUefiOutputSupport (
+  VOID
+  )
+{
+  EFI_STATUS  Status;
+
+  Status = OcProvideConsoleGop (FALSE);
+  if (EFI_ERROR (Status)) {
+    LogStage ('2');
+    return Status;
+  }
+
+  LogStage ('3');
+
+  OcSetConsoleResolution (
+    0,
+    0,
+    0,
+    FALSE
+    );
+
+  if (FeaturePcdGet (PcdEnableGopDirect)) {
+    OcUseDirectGop (-1);
+  }
+
+  OcSetupConsole (
+    OcConsoleRendererBuiltinGraphics,
+    FALSE,
+    FALSE,
+    FALSE,
+    FALSE
+    );
+
+  return EFI_SUCCESS;
+}
+
+STATIC
+EFI_STATUS
+ProvideGop (
+  VOID
+  )
+{
+  LogStage ('1');
+
+  OcUnlockAppleFirmwareUI ();
+
+  OcForgeUefiSupport (TRUE, TRUE);
+  OcReloadOptionRoms ();
+
+  BdsLibConnectAllDriversToAllControllers ();
+
+  return LoadUefiOutputSupport ();
+}
+
+STATIC
+EFI_STATUS
+SetDefaultVariable (
+  CHAR16    *Name,
+  EFI_GUID  *Guid,
+  UINTN     Size,
+  VOID      *Data
+  )
+{
+  EFI_STATUS  Status;
+  UINTN       DataSize;
+
+  DataSize = 0;
+  Status   = gRT->GetVariable (
+                    Name,
+                    Guid,
+                    NULL,
+                    &DataSize,
+                    NULL
+                    );
+
+  //
+  // Do not modify existing value.
+  //
+  if (Status == EFI_BUFFER_TOO_SMALL) {
```

Comment:
What if there is a different error? Imo check for "not success" (i.e., `!= EFI_NOT_FOUND` I think) rather than an explicit "failure" (i.e., var does exist and can be read successfully).

Reply:
I'm not convinced I agree. My logic, fwiw, is that I might as well attempt the write on anything other than "the variable exists already and can be read", no?

Reply:
Also - I'll have to retest the logic of detecting whether we are injected or running as Driver#### if I modify this...!

Reply:
Well, if that makes sense to you, sure, I have no context how “bad” it is to have a false positive over a false negative.